lib/Sema/TypeCheckAttr.cpp

```diff
@@ -2171,16 +2149,85 @@ SynthesizeMainFunctionRequest::evaluate(Evaluator &evaluator,
   // mainType.main() from the entry point, and that would require fully
   // type-checking the call to mainType.main().
 
-  constraints::ConstraintSystemOptions lookupOptions;
-  if (context.LangOpts.EnableAsyncMainResolution)
-    lookupOptions |=
-        constraints::ConstraintSystemFlags::ConsiderNominalTypeContextsAsync;
+  constraints::ConstraintSystem CS(
+      declContext, constraints::ConstraintSystemFlags::IgnoreAsyncSyncMismatch);
+  constraints::ConstraintLocator *locator = CS.getConstraintLocator({});
+  // Allowed main function types
+  // `() -> Void`
+  // `() async -> Void`
+  // `() throws -> Void`
+  // `() async throws -> Void`
+  // `@MainActor () -> Void`
+  // `@MainActor () async -> Void`
+  // `@MainActor () throws -> Void`
+  // `@MainActor () async throws -> Void`
+  {
+    llvm::SmallVector<Type, 8> mainTypes = {
+
+        FunctionType::get(/*params*/ {}, context.TheEmptyTupleType,
+                          ASTExtInfo()),
+        FunctionType::get(
+            /*params*/ {}, context.TheEmptyTupleType,
+            ASTExtInfoBuilder().withAsync().build()),
+
+        FunctionType::get(/*params*/ {}, context.TheEmptyTupleType,
+                          ASTExtInfoBuilder().withThrows().build()),
+
+        FunctionType::get(
+            /*params*/ {}, context.TheEmptyTupleType,
+            ASTExtInfoBuilder().withAsync().withThrows().build())};
+
+    Type mainActor = context.getMainActorType();
+    if (mainActor) {
+      mainTypes.push_back(FunctionType::get(
+          /*params*/ {}, context.TheEmptyTupleType,
+          ASTExtInfoBuilder().withGlobalActor(mainActor).build()));
+      mainTypes.push_back(FunctionType::get(
+          /*params*/ {}, context.TheEmptyTupleType,
+          ASTExtInfoBuilder().withAsync().withGlobalActor(mainActor).build()));
+      mainTypes.push_back(FunctionType::get(
+          /*params*/ {}, context.TheEmptyTupleType,
+          ASTExtInfoBuilder().withThrows().withGlobalActor(mainActor).build()));
+      mainTypes.push_back(FunctionType::get(/*params*/ {},
+                                            context.TheEmptyTupleType,
+                                            ASTExtInfoBuilder()
+                                                .withAsync()
+                                                .withThrows()
+                                                .withGlobalActor(mainActor)
+                                                .build()));
+    }
+
+    llvm::SmallVector<constraints::Constraint *, 4> mainTypeConstraints;
+    for (const Type &mainType : mainTypes) {
+      constraints::Constraint *fnConstraint =
+          constraints::Constraint::createMember(
+              CS, constraints::ConstraintKind::ValueMember,
+              nominal->getInterfaceType(), mainType,
```

Comment:
If one of the choices produced by member lookup is `() async -> Void` and the current main type is `() -> Void` it would just get accepted as if member was `() -> Void` and that's what we want? If I remember it correctly we want to make sure that there is absolutely no other suitable choices anywhere in the supertype tree of the base type? If so a better approach might be to allocate a type variable to record all the correct types as equal to that type variable and only do member lookup once e.g.

```
$T3 - member type

disjunction [$T3 equal () -> Void, $T3 equal () async -> Void, ....]
<base> member `main` -> $T3
```

That way the solver would do some of the work here.

Reply:
Alright, implemented. Yeah, this really reduces the number of scopes it explores.
For the `Config3` ambiguous test case, the original explored 41 scopes. With the type variable equality disjunction, we only explore 21 scopes.